### PR TITLE
occlusion query

### PR DESCRIPTION
**Pull request by stanleyljl**
Re: cocos-creator/3d-tasks#https://github.com/cocos-creator/3d-tasks/issues/9047

Changelog:
 * add occlusion query

### CODE REVIEW

File: cocos/3d/models/baked-skinning-model.ts [hunk 1]
```diff
@@ -229,6 +229,10 @@ export class BakedSkinningModel extends MorphModel {
         }
     }
 
+    public _updateLocalWorldBoundDescriptors (subModelIndex: number, descriptorSet: DescriptorSet) {
+        super._updateLocalWorldBoundDescriptors(subModelIndex, descriptorSet);
```
Comment (pandamicro) [inline cocos/3d/models/baked-skinning-model.ts:233]: It doesn't seem meaningful to overwrite this function

File: cocos/core/pipeline/pipeline-scene-data.ts [hunk 2]
```diff
@@ -112,9 +120,39 @@ export class PipelineSceneData {
     public activate (device: Device, pipeline: RenderPipeline) {
         this._device = device;
         this._pipeline = pipeline;
+
+        this.initOcclusionQuery();
```
Comment (pandamicro) [inline cocos/core/pipeline/pipeline-scene-data.ts:124]: Please only initialize if device support occlusion query

File: cocos/core/pipeline/pipeline-scene-data.ts [hunk 3]
```diff
@@ -123,8 +161,47 @@ export class PipelineSceneData {
...
+        this._occlusionQueryVertexBuffer?.destroy();
+        this._occlusionQueryVertexBuffer = null;
+        this._occlusionQueryIndicesBuffer?.destroy();
+        this._occlusionQueryIndicesBuffer = null;
         if (JSB) {
             this._nativeObj = null;
         }
     }
+
+    private createInputAssembler () {
```
Comment (pandamicro) [inline cocos/core/pipeline/pipeline-scene-data.ts:175]: please add `_` as prefix of private function name. Also the function name is not meaningful, if it's for creating occlusion query assembler, you can use name like `_createOcclusionQueryIA`

File: cocos/core/pipeline/pipeline-scene-data.ts [hunk 4]
```diff
@@ -123,8 +161,47 @@ export class PipelineSceneData {
         this.skybox.destroy();
         this.fog.destroy();
         this.shadows.destroy();
+        this._occlusionQueryInputAssembler?.destroy();
+        this._occlusionQueryInputAssembler = null;
+        this._occlusionQueryVertexBuffer?.destroy();
+        this._occlusionQueryVertexBuffer = null;
+        this._occlusionQueryIndicesBuffer?.destroy();
+        this._occlusionQueryIndicesBuffer = null;
```
Comment (pandamicro) [inline cocos/core/pipeline/pipeline-scene-data.ts:169]: check whether occlusion query is initialized

File: cocos/core/renderer/scene/model.ts [hunk 5]
```diff
@@ -209,6 +213,9 @@ export class Model {
     private _lightmap: Texture2D | null = null;
     private _lightmapUVParam: Vec4 = new Vec4();
 
+    protected _localWorldBoundData = new Float32Array(UBOLocalWorldBound.COUNT);
```
Comment (pandamicro) [inline cocos/core/renderer/scene/model.ts:216]: It's better to only initialize it when needed

File: cocos/core/renderer/scene/model.ts [hunk 6]
```diff
@@ -360,6 +383,25 @@ export class Model {
             this._localBuffer.update(this._localData);
             this._applyLocalData();
             this._applyLocalBuffer();
+
+            this.updateWorldBoundUBOs();
+        }
+    }
+
+    public updateWorldBoundUBOs () {
```
Comment (pandamicro) [inline cocos/core/renderer/scene/model.ts:391]: Does  it need to be public, seems only used internally

File: cocos/core/renderer/scene/model.ts [hunk 7]
```diff
@@ -114,6 +114,10 @@ export class Model {
         return this._localBuffer;
     }
 
+    get localWorldBoundBuffer () {
```
Comment (pandamicro) [inline cocos/core/renderer/scene/model.ts:117]: The name `local world bound` is very strange, maybe name it as `model world bound`

And ay public property of function need a API description

**Review by pandamicro — CHANGES_REQUESTED** [review]

File: cocos/3d/models/baked-skinning-model.ts (hunk 1 above)
Comment (stanleyljl) [reply cocos/3d/models/baked-skinning-model.ts:233]: 不这么写构建失败

File: cocos/core/pipeline/pipeline-scene-data.ts (hunk 2 above)
Comment (stanleyljl) [reply cocos/core/pipeline/pipeline-scene-data.ts:124]: 估计还是得写上，因为支持动态开关，如果实时销毁/创建流程会很乱。不过这数据量很小，一个model多两个vector4

File: cocos/core/pipeline/pipeline-scene-data.ts (hunk 3 above)
Comment (stanleyljl) [reply cocos/core/pipeline/pipeline-scene-data.ts:175]: 1

File: cocos/core/pipeline/pipeline-scene-data.ts (hunk 4 above)
Comment (stanleyljl) [reply cocos/core/pipeline/pipeline-scene-data.ts:169]: 必然会创建

File: cocos/core/renderer/scene/model.ts (hunk 5 above)
Comment (stanleyljl) [reply cocos/core/renderer/scene/model.ts:216]: 同上

File: cocos/core/renderer/scene/model.ts (hunk 6 above)
Comment (stanleyljl) [reply cocos/core/renderer/scene/model.ts:391]: 1

**Pull request closed without merging**